File: 2699-count-the-number-of-fair-pairs/2699-count-the-number-of-fair-pairs.cpp
class Solution {
public:

//  TC -> O(nlog(n))
    long long countFairPairs(vector<int>& nums, int lower, int upper) {
        long long fairPairs = 0;
        int n = nums.size();

        sort(begin(nums), end(nums)); // O(nlog(n))

        for(int i = 0; i<n; i++){ // O(n)
            int idx = lower_bound(begin(nums)+i+1, end(nums), lower - nums[i])-begin(nums); // O(long(n))
            int numLessThanLower = idx - i -1;

            idx = upper_bound(begin(nums) + i + 1, end(nums), upper-nums[i])- begin(nums); // O(log(n))

            int numLessThanOrEqualToUpper = idx - i -1;

            fairPairs += numLessThanOrEqualToUpper - numLessThanLower;
        }

        return fairPairs;
    }

};

File: 2699-count-the-number-of-fair-pairs/2699-count-the-number-of-fair-pairs.java
class Solution {
    // TC : O(nlog(n))
    public long countFairPairs(int[] nums, int lower, int upper) {
        long fairPairs = 0;
        int n = nums.length;

        Arrays.sort(nums); // O(nlog(n))

        int left = 0, right = n-1;
        long countPairLessThanUpper = 0;
        while(left < right){ // O(log(n))
            if(nums[left] + nums[right] <= upper){
                countPairLessThanUpper += (right - left);
                left++;
            }else{
                right--;
            }
        }
        left = 0; right = n-1;
        long lowerPairCount = 0;
        while(left < right){ // O(log(n))
            if(nums[left] + nums[right] < lower){
                lowerPairCount += (right - left);
                left++;
            }else{
                right--;
            }
        }
        fairPairs = countPairLessThanUpper - lowerPairCount;
        return fairPairs;
    }
}